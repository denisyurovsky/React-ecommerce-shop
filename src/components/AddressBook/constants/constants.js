export const LISTBOX_PADDING = 8;
export const DOUBLE_LISTBOX_PADDING = LISTBOX_PADDING * 2;
export const LIMIT_NUMBER_ITEMS = 8;
export const ITEM_SIZE_LG = 48;
export const ITEM_SIZE_MD = 36;
