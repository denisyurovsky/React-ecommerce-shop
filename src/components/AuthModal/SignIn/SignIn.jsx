import {
  Checkbox,
  FormControlLabel,
  TextField,
  Container,
  Button,
} from '@mui/material';
import PropTypes from 'prop-types';
import React, { useState } from 'react';

import {
  ERROR,
  MIN_PASSWORD_LENGTH,
  authStatus,
} from '../../../helpers/constants/authConstants';
import CustomLoadingButton from '../../ui-kit/buttons/CustomLoadingButton';
import PasswordForm from '../PasswordForm';
import useErrorHandler from '../useErrorHandler';

const { LOCKED, PENDING, FULFILLED, REJECTED, IDLE } = authStatus;

const SignIn = ({ sendForm, status, errorMessage }) => {
  const isLoading = status === authStatus.PENDING;
  const isLocked = status === authStatus.LOCKED;
  const isFailed = status === authStatus.REJECTED;
  const disabled = isLocked || isLoading;

  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [isKeep, setIsKeep] = useState(false);
  const [isFinished, setIsFinished] = useState(false);
  const { isError, resetError } = useErrorHandler(isFailed || isLocked);

  const handleEmailChange = (e) => {
    const input = e.target.value;

    setEmail(input);
    resetError();
    setIsFinished(checkIsFinished(input, password));
  };

  const handlePasswordChange = (e) => {
    const input = e.target.value;

    setPassword(input);
    resetError();
    setIsFinished(checkIsFinished(email, password));
  };

  const handleKeepChange = () => setIsKeep(!isKeep);

  const checkEmail = (email) => email.match(/@/);
  const checkPassword = (password) => password.length > MIN_PASSWORD_LENGTH;
  const checkIsFinished = (email, password) =>
    checkEmail(email) && checkPassword(password);

  const prepareToSend = () => sendForm({ password, email });

  return (
    <Container>
      <TextField
        error={isError}
        helperText={isError ? errorMessage : null}
        autoComplete="off"
        fullWidth
        variant="standard"
        margin="normal"
        label="Email"
        value={email}
        onChange={handleEmailChange}
        disabled={disabled}
      />
      <PasswordForm
        isError={isError}
        errorMessage={errorMessage}
        password={password}
        onChange={handlePasswordChange}
        disabled={disabled}
      />
      <FormControlLabel
        sx={{ mb: 1 }}
        control={<Checkbox onChange={handleKeepChange} checked={isKeep} />}
        label="Remember me"
        disabled={disabled}
      />
      {isLoading ? (
        <CustomLoadingButton label="Loading..." isFullWidth={true} />
      ) : (
        <Button
          fullWidth
          variant="contained"
          onClick={prepareToSend}
          disabled={!isFinished || isLocked}
        >
          Sign in
        </Button>
      )}
    </Container>
  );
};

SignIn.defaulProps = {
  errorMessage: ERROR.LOGIN,
  status: IDLE,
};

SignIn.propTypes = {
  sendForm: PropTypes.func.isRequired,
  status: PropTypes.oneOf([PENDING, FULFILLED, REJECTED, IDLE, LOCKED]),
  errorMessage: PropTypes.string,
};

export default SignIn;
