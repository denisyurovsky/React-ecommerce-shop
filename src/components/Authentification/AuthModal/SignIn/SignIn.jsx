import { Checkbox, FormControlLabel, Box, Button } from '@mui/material';
import PropTypes from 'prop-types';
import React, { useState } from 'react';

import {
  MIN_PASSWORD_LENGTH,
  authStatus,
} from '../../../../constants/authConstants';
import LoadingButton from '../../../ui-kit/buttons/LoadingButton';
import { StandardTextField } from '../../../ui-kit/StandardTextField/StandardTextField';
import PasswordForm from '../PasswordForm';
import useErrorHandler from '../useErrorHandler';

const { LOCKED, PENDING, FULFILLED, REJECTED, IDLE } = authStatus;

const SignIn = ({ sendForm, status, errorMessage }) => {
  const isLoading = status === PENDING;
  const isLocked = status === LOCKED;
  const isFailed = status === REJECTED;
  const disabled = isLocked || isLoading;

  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [isKeep, setIsKeep] = useState(false);
  const [isFinished, setIsFinished] = useState(false);
  const { isError, resetError } = useErrorHandler(isFailed || isLocked);

  const handleEmailChange = (e) => {
    const input = e.target.value;

    setEmail(input);
    resetError();
    setIsFinished(checkIsFinished(input, password));
  };

  const handlePasswordChange = (e) => {
    const input = e.target.value;

    setPassword(input);
    resetError();

    e.target.hasAttribute('autocompleted')
      ? setIsFinished(true)
      : setIsFinished(checkIsFinished(email, password));
  };

  const handleKeepChange = () => setIsKeep(!isKeep);

  const checkEmail = (email) => email.match(/@/);
  const checkPassword = (password) => password.length > MIN_PASSWORD_LENGTH;
  const checkIsFinished = (email, password) =>
    checkEmail(email) && checkPassword(password);

  const prepareToSend = () => sendForm({ password, email, isKeep });

  return (
    <Box>
      <StandardTextField
        onChange={handleEmailChange}
        disabled={disabled}
        autoComplete="off"
        error={isError}
        helperText={isError ? errorMessage : null}
        labelText="Email"
        value={email}
      />
      <PasswordForm
        isError={isError}
        errorMessage={errorMessage}
        password={password}
        onChange={handlePasswordChange}
        disabled={disabled}
      />
      <FormControlLabel
        sx={{ mb: 1 }}
        control={<Checkbox onChange={handleKeepChange} checked={isKeep} />}
        label="Remember me"
        disabled={disabled}
      />

      {isLoading ? (
        <LoadingButton label="Loading..." isFullWidth={true} />
      ) : (
        <Button
          fullWidth
          variant="contained"
          onClick={prepareToSend}
          disabled={!isFinished || isLocked}
        >
          Sign in
        </Button>
      )}
    </Box>
  );
};

SignIn.defaultProps = {
  errorMessage: '',
  status: IDLE,
};

SignIn.propTypes = {
  sendForm: PropTypes.func.isRequired,
  status: PropTypes.oneOf([PENDING, FULFILLED, REJECTED, IDLE, LOCKED]),
  errorMessage: PropTypes.string,
};

export default SignIn;
