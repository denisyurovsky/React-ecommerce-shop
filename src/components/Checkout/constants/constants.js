export const PANEL = {
  DELIVERY_ADDRESS: 'Delivery Address',
  PERSONAL_INFORMATION: 'Personal Information',
  PERSONAL_ADDRESS: 'Personal Address',
  PAYMENT_METHOD: 'Payment method',
  ORDER_CONFIRMATION: 'Order confirmation',
};
