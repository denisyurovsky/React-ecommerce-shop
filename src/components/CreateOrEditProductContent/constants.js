export const PRICE_ERROR = 'Only numbers';
export const DESCRIPTION_ERROR = 'Only latin characters';
export const LENGTH_ERROR = 'Should contain at least one symbol';
export const DISCOUNT_ERROR = 'Discount should be less than price';
export const SELECTION_ERROR = 'Choose category';
