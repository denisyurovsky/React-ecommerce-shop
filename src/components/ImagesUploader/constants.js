export const LOSS_OF_IMAGES = 'We lost images';

export const STATUS = {
  CROPPED: 'cropped',
  CROPPING: 'cropping',
  IDLE: 'idle',
};

export const TEST_IMG_URL = 'http://someURL';
