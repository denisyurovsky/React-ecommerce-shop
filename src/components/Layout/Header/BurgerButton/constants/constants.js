import { pathNames } from '../../../../../constants/pathNames';

export const LINKS = [
  {
    name: 'HOME',
    link: '/',
  },
  {
    name: 'PRODUCTS',
    link: pathNames.PRODUCTS,
  },
];
