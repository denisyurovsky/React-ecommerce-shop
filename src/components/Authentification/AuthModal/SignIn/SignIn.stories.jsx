import React from 'react';

import { authStatus, ERROR } from '../../../../constants/authConstants';

import SignIn from './SignIn';

const { IDLE, LOCKED, PENDING, REJECTED, FULFILLED } = authStatus;

export default {
  title: 'AuthModal/SignIn',
  component: SignIn,
  argTypes: { sendForm: { action: 'Sending form' } },
};

const Template = (args) => <SignIn {...args} />;

export const Default = Template.bind({});
Default.args = {
  status: IDLE,
  errorMessage: '',
};

export const Success = Template.bind({});
Success.args = {
  status: FULFILLED,
  errorMessage: '',
};

export const Loading = Template.bind({});
Loading.args = {
  status: PENDING,
  errorMessage: '',
};

export const Wrong = Template.bind({});
Wrong.storyName = 'Wrong credentials';
Wrong.args = {
  status: REJECTED,
  errorMessage: ERROR.LOGIN,
};

export const Locked = Template.bind({});
Locked.storyName = 'Many attempts';
Locked.args = {
  status: LOCKED,
  errorMessage: ERROR.LOCK,
};
