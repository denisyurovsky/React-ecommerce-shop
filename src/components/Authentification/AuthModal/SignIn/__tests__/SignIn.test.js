import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import { authStatus } from '../../../../../constants/authConstants';
import SignIn from '../SignIn';

const { REJECTED, PENDING, LOCKED } = authStatus;

const fillForms = (emailValue, passwordValue, isKeep) => {
  const btn = screen.getByRole('button');
  const email = screen.getByLabelText('Email');
  const password = screen.getByLabelText('Password');
  const rememberMeCheck = screen.getByLabelText('Remember me');

  userEvent.type(email, emailValue);
  userEvent.type(password, passwordValue);
  if (isKeep) {
    userEvent.click(rememberMeCheck);
  }

  return { btn, email, password, rememberMeCheck };
};

const getInputs = () => {
  const { btn, email, password } = fillForms('look@', 'theP', false);
  const rememberMeCheck = screen.getByLabelText('Remember me');
  const visibilityCheck = screen
    .getByTestId('visibility')
    .querySelector('input[type="checkbox"]');
  const arrayOfInputs = [
    btn,
    email,
    password,
    rememberMeCheck,
    visibilityCheck,
  ];

  return {
    btn,
    email,
    password,
    visibilityCheck,
    rememberMeCheck,
    arrayOfInputs,
  };
};

const errorMessage = 'Testing error message...';

describe('SignIn snapshot:', () => {
  it('should render a valid snapshot', () => {
    const { asFragment } = render(<SignIn sendForm={() => {}} />);

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Send form button tests:', () => {
  const callback = jest.fn();

  beforeEach(() => {
    render(<SignIn sendForm={callback} />);
  });

  it('should have active state with checked remember me', () => {
    const login = 'ivan@ya';
    const passwordValue = 'pass';
    const { btn, email, password, rememberMeCheck } = fillForms(
      login,
      passwordValue,
      true
    );

    expect(email.value).toBe(login);
    expect(password.value).toBe(passwordValue);
    expect(rememberMeCheck.checked).toBe(true);
    expect(btn.disabled).toBe(false);
  });

  it('should have active state with unchecked remember me', () => {
    const login = 'ivan@ya';
    const passwordValue = 'pass';
    const { btn, email, password, rememberMeCheck } = fillForms(
      login,
      passwordValue,
      false
    );

    expect(email.value).toBe(login);
    expect(password.value).toBe(passwordValue);
    expect(rememberMeCheck.checked).toBe(false);
    expect(btn.disabled).toBe(false);
  });

  it('should send form with correct credentials with checked remember me', () => {
    const email = 'petr@';
    const password = 'weak';
    const isKeep = true;
    const { btn } = fillForms(email, password, isKeep);

    fireEvent.click(btn);
    expect(callback).toBeCalledWith({ email, password, isKeep });
  });

  it('should send form with correct credentials with unchecked remember me', () => {
    const email = 'petr@';
    const password = 'weak';
    const isKeep = false;
    const { btn } = fillForms(email, password, isKeep);

    fireEvent.click(btn);
    expect(callback).toBeCalledWith({ email, password, isKeep });
  });
});

describe('Test lock state', () => {
  beforeEach(() => {
    render(
      <SignIn sendForm={() => {}} status={LOCKED} errorMessage={errorMessage} />
    );
  });

  it('should lock all inputs', () => {
    const { arrayOfInputs } = getInputs();

    arrayOfInputs.forEach((el) => {
      expect(el.disabled).toBe(true);
    });
  });

  it('should display error message', () => {
    expect(screen.getAllByText(errorMessage)).toHaveLength(2);
  });
});

describe('Test loading state', () => {
  beforeEach(() => {
    render(
      <SignIn
        sendForm={() => {}}
        status={PENDING}
        errorMessage={errorMessage}
      />
    );
  });

  it('should lock all inputs', () => {
    const { arrayOfInputs } = getInputs();

    arrayOfInputs.forEach((el) => {
      expect(el.disabled).toBe(true);
    });
  });

  it('should change button text', () => {
    expect(screen.getByRole('button')).toHaveTextContent(/loading/i);
  });

  it('should not display error', () => {
    expect(screen.queryAllByText(errorMessage)).toHaveLength(0);
  });
});

describe('Test error state', () => {
  beforeEach(() => {
    render(
      <SignIn
        sendForm={() => {}}
        errorMessage={errorMessage}
        status={REJECTED}
      />
    );
  });

  it('should remove error message after user types something', () => {
    const { email } = getInputs();

    userEvent.type(email, 'k');
    expect(screen.queryAllByText(errorMessage)).toHaveLength(0);
  });

  describe('snapshot', () => {
    it('should show error and do not disable inputs', () => {
      const { container } = render(
        <SignIn
          sendForm={() => {}}
          errorMessage={errorMessage}
          status={REJECTED}
        />
      );

      expect(container).toMatchSnapshot();
    });
  });
});

describe('Test base state', () => {
  beforeEach(() => {
    render(<SignIn sendForm={() => {}} />);
  });

  it('Checkbox remember me should work', () => {
    const rememberMeCheck = screen.getByLabelText('Remember me');

    expect(rememberMeCheck.checked).toBe(false);
    fireEvent.click(rememberMeCheck);
    expect(rememberMeCheck.checked).toBe(true);
  });
});
