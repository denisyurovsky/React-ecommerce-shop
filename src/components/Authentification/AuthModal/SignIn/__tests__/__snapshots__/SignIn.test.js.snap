// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn snapshot: should render a valid snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="mui-1"
        id="mui-1-label"
      >
        Email
      </label>
      <div
        class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-ghsjzk-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          autocomplete="off"
          class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          id="mui-1"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="mui-2"
        id="mui-2-label"
      >
        Password
      </label>
      <div
        class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-ghsjzk-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          autocomplete="off"
          class="MuiInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-1x51dt5-MuiInputBase-input-MuiInput-input"
          id="mui-2"
          type="password"
          value=""
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        >
          <span
            class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
            data-testid="visibility"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="VisibilityOffIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78 3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
        </div>
      </div>
    </div>
    <label
      class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-1apuumd-MuiFormControlLabel-root"
    >
      <span
        class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
      >
        <input
          class="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate="false"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckBoxOutlineBlankIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
      >
        Remember me
      </span>
    </label>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root Mui-disabled  css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
      disabled=""
      tabindex="-1"
      type="button"
    >
      Sign in
    </button>
  </div>
</DocumentFragment>
`;

exports[`Test error state snapshot should show error and do not disable inputs 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary Mui-error css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="mui-28"
        id="mui-28-label"
      >
        Email
      </label>
      <div
        class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl css-ghsjzk-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-describedby="mui-28-helper-text"
          aria-invalid="true"
          autocomplete="off"
          class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          id="mui-28"
          type="text"
          value=""
        />
      </div>
      <p
        class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium css-1d1r5q-MuiFormHelperText-root"
        id="mui-28-helper-text"
      >
        Testing error message...
      </p>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary Mui-error css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="mui-29"
        id="mui-29-label"
      >
        Password
      </label>
      <div
        class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-ghsjzk-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-describedby="mui-29-helper-text"
          aria-invalid="true"
          autocomplete="off"
          class="MuiInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-1x51dt5-MuiInputBase-input-MuiInput-input"
          id="mui-29"
          type="password"
          value=""
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        >
          <span
            class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
            data-testid="visibility"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="VisibilityOffIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78 3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
        </div>
      </div>
      <p
        class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium css-1d1r5q-MuiFormHelperText-root"
        id="mui-29-helper-text"
      >
        Testing error message...
      </p>
    </div>
    <label
      class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-1apuumd-MuiFormControlLabel-root"
    >
      <span
        class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
      >
        <input
          class="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate="false"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckBoxOutlineBlankIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
      >
        Remember me
      </span>
    </label>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root Mui-disabled  css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
      disabled=""
      tabindex="-1"
      type="button"
    >
      Sign in
    </button>
  </div>
</div>
`;
